### ISSUE TITLE
Issue with dynamically required files

### ISSUE THREAD

**Issue by mdcarter** [zeit/ncc#74]
I'm using a custom Hapi made server and I use this kind of require a lot (for loading plugins, methods and route : 
```await server.register(require(filepath))```

I use this within loops that read folders and files to load everything dynamically.
It's not working (using run or build, either way), the required object/library is not available.

**Comment by guybedford** [comment]
There may be some things we can do here. It would help a lot of you could share how `filepath` is initially constructed and passed into the registration function in this example? Or is it entirely a user-API?

**Comment by mdcarter** [comment]
Of course, no problems.

Here is a sample code used to dynamically loads server methods and plugins to the Hapi server object : 
```
await Promise.all([
  glob.sync(`./methods/**/index.js`).map(async filepath => {
    let method = require(filepath)
    await server.method(method.name, method.register)
  }),
  glob.sync(`./plugins/**/index.js`).map(async filepath => {
    await server.register(require(filepath))
  }),
])
```

**Comment by guybedford** [comment]
So this is a tricky case to apply static analysis to, as it requires glob-based knowledge for the static analysis. That is something that could be done eventually though, either directly, or through fs virtualization approaches that may be explored in future.

If the above were written to use something like `require(path.resolve('./methods/' + relpath))` then that is something we can statically analyze right now.

**Comment by mdcarter** [comment]
@guybedford It's seems to be working fine the way you told me, thank you very much !
Now I juste have to wrap my head around exactly what's happening/if it's interesting for my project.
Thank you again !

**Comment by brettdh** [comment]
@guybedford I am having similar problems, but with third-party code:

https://spectrum.chat/zeit/now/error-with-require-a-particular-package-cannot-find-module-var-task-package-json~9fdea285-e149-4fdc-ad54-3e4eced7dc86

I may be able to apply a `path.resolve` workaround as you suggested (though I'd welcome a little more detail as to why that would work). Perhaps consider re-opening this to track future support for dynamic requires?

Perhaps also consider detecting such situations and warning/erroring? This took a while to track down, and an informative error message about this limitation would be really helpful.

**Comment by brettdh** [comment]
The relevant code: https://github.com/holidayextras/jsonapi-server/blob/master/lib/routes/index.js

**Comment by KevinWang15** [comment]
let's try `eval("require")(filepath)` ... 😓

**Comment by drenel18** [comment]
@KevinWang15  Wow!!! How do you come up with it? 😀

**Comment by guybedford** [comment]
Note `__non_webpack_require__` works in ncc since it is just webpack.

**Comment by KevinWang15** [comment]
seems like `__non_webpack_require__`  is compiled to exactly `eval("require")` by webpack 😀

**Comment by guybedford** [comment]
Yes, that's what we do in ncc internally :)

**Comment by Vadorequest** [comment]
FYI just read https://hackernoon.com/building-isomorphic-javascript-packages-1ba1c7e558c5 and adding ` yarn add -D @types/webpack-env` fixed my TypeScript compilation failure when using `__non_webpack_require__`.

**Comment by Songkeys** [comment]
@guybedford 
> So this is a tricky case to apply static analysis to, as it requires glob-based knowledge for the static analysis. That is something that could be done eventually though, either directly, **or through fs virtualization approaches** that may be explored in future.

Hi, this is a old thread but could you please be more specific about the vfs approaches? How would that work? I ran into a similar problem (a [framework](https://eggjs.org) that loads js files dynamically according to the project structure) and tried to sovle it from ncc itself but failed.

(I've checked the vercel/webpack-asset-relocator-loader package but it seems that it recognises the js files as assets and relocating them without compiling them, which is not what I want. I have no clue about what to do now.)

Thank you.

**Comment by guybedford** [comment]
@Songkeys each case is very specific here, so if you have a code example of why eggjs is not properly bundling that would help to provide suggestions further.

**Comment by Songkeys** [comment]
@guybedford 
I think the issue is still about dynamic requiring.

```js
require('xxx') // this will be recognised

const dynamicRequire = (packageName) => require(packageName)
dynamicRequire('xxx') // this won't be recognised
```

**Comment by KevinWang15** [comment]
PoC of file system virtualization


### bundler
```js
const fs = require("fs");
const path = require("path");
const {GlobSync} = require("glob");
process.chdir("./target");


const requireRedefined = dirName => `function require(file) {
    const module={};
    return eval(fileMap['./' + path.relative(process.cwd(), path.resolve('${dirName}/'+file))] + 'module.exports;');
}
`;

const fileMap = {};
const files = GlobSync("./**/*.js").found;
for (let file of files) {
    fileMap[file] = requireRedefined(path.dirname(file)) + fs.readFileSync(file, {encoding: "UTF-8"});
}

const resultFile = [
    `const path = require('path');`,
    `const fileMap = ${JSON.stringify(fileMap, null, 4)};`,
    `eval(fileMap["./index.js"]);`
].join("\n");

process.stdout.write(resultFile);
```


### input
```
./index.js
["a.js", "b.js"].forEach(file => {
    console.log(require("./dependencies/" + file));
});

./dependencies/a.js
const a1 = require("./dependencies/a1.js")
module.exports = "AAA" + a1;

./dependencies/dependencies/a1.js
module.exports = "A1";

./dependencies/b.js
module.exports = "BBB";
```

### output
```js
const path = require('path');
const fileMap = {
    "./dependencies/a.js": "function require(file) {\n    const module={};\n    return eval(fileMap['./' + path.relative(process.cwd(), path.resolve('./dependencies/'+file))] + 'module.exports;');\n}\nconst a1 = require(\"./dependencies/a1.js\")\nmodule.exports = \"AAA\" + a1;",
    "./dependencies/b.js": "function require(file) {\n    const module={};\n    return eval(fileMap['./' + path.relative(process.cwd(), path.resolve('./dependencies/'+file))] + 'module.exports;');\n}\nmodule.exports = \"BBB\";",
    "./dependencies/dependencies/a1.js": "function require(file) {\n    const module={};\n    return eval(fileMap['./' + path.relative(process.cwd(), path.resolve('./dependencies/dependencies/'+file))] + 'module.exports;');\n}\nmodule.exports = \"A1\";",
    "./index.js": "function require(file) {\n    const module={};\n    return eval(fileMap['./' + path.relative(process.cwd(), path.resolve('./'+file))] + 'module.exports;');\n}\n[\"a.js\", \"b.js\"].forEach(file => {\n    console.log(require(\"./dependencies/\" + file));\n});\n"
};
eval(fileMap["./index.js"]);
```


an interesting concept worth exploring.. @Songkeys

**Comment by guybedford** [comment]
Thanks for the clear reports. The problem in the eggjs case is dynamic require detection over virtualization it looks like to me. I believe some of these dynamic require cases are handled better by the Webpack upgrade in https://github.com/vercel/ncc/pull/579 since Webpack improves the dynamic require detection handling in one of those releases. I'd suggest trying that out once it lands then reposting a new issue if it still isn't working.

**Comment by Songkeys** [comment]
@guybedford  Thank you. I've tried the latest version (0.24.0) from #579. It handles dynamic requiring better but not perfectly. Anyway, think the issue #578 can be closed now.

The framework I'm using check a folder/files on runtime. For example, the project structure is like:

```
- app/
   - controller/
   - router.js
   - index.js
```

There is no direct `require` statement to import `controller` directory or `router.js` into the framework. The framework checks if there exists such a file/folder on runtime (e.g. `path.resolve(process.cwd(), 'controller')`) and then requires them. Since the project structure is destroyed by ncc into a single file, it never gonna find the expected files.

I don't think this case could be handled by static analyzation. The PoC of file system virtualization proposed by @KevinWang15  is more like a solution to me.

Let me know if you need more detail or a new issue.

**Comment by devthejo** [comment]
here is my PoC of a 2 pass working solution:

each dynamic require have to run one time before ncc build
won't work with conditional runtime requires, but useful for files structure based tools

index.js
```js
const path = require("path")
const fs = require("fs")

const dynamicFilename = "./build/requires.js"
const dynamicRequires = new Set()
const dynamicRequireRegister = (src)=>{
  if (dynamicRequires.has(src)){
    return
  }
  dynamicRequires.add(src)
  const exports = [...dynamicRequires].reduce((acc, req)=>{
    const target = !req.startsWith(".") ? req : path.join(process.cwd(), req)
    
    acc.push(`${JSON.stringify(req)}: require(${JSON.stringify(target)})`)
    return acc
  },[])
  fs.mkdirSync(path.dirname(dynamicFilename),{recursive:true})
  fs.writeFileSync(dynamicFilename, `module.exports={${exports.join(",")}}`)
}
const dynamicRequire = (r)=>{
  let required
  try {
    // build time, each dynamic require have to run one time before ncc build
    // won't work with conditional runtime requires, but useful for files structure based tools
    require(r)
    required = true
  }
  catch(_e){
    // do nothing
  }
  if (required){
    dynamicRequireRegister(r)
  }
  // the require target has to be static, dont replace it with variable dynamicFilename
  return require("./build/requires")[r]
}

const r = "./my-dynamic-require.js"

const hello = dynamicRequire(r)

if(process.env.BUILD){
  return
}

hello("folks")
```

my-dynamic-require
```js
module.exports = (some) => console.log(`Hello ${some} !`)
```

then call
```sh
BUILD=1 node index.js && ncc build index.js
```

and you're good ;)
```sh
node dist/index.js
```

**Comment by Songkeys** [comment]
@devthejo i think we share the same thought. #734 guybedford has given his suggestion on this idea.
